### DISCUSSION TITLE
[@types/sshpk] Inconsistency in PrivateKey.createSign(AlgorithmType)

### DISCUSSION THREAD

**Discussion by OlivierMartineau** [DefinitelyTyped/DefinitelyTyped#58887] [Issues with a @types package]
In `PrivateKey` class, `createSign()` [function](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e6467b3284fe603dd0a2c21f89ac70a7c61d45c9/types/sshpk/index.d.ts#L712
) is said to require an `AlgorithmType` object
```js
class PrivateKey {
    createSign(hashAlgo: AlgorithmType): Signer;
}
```

This type is [defined](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/e6467b3284fe603dd0a2c21f89ac70a7c61d45c9/types/sshpk/index.d.ts#L19) as
```js
type AlgorithmType = 'dsa' | 'rsa' | 'ecdsa' | 'ed25519';
```

As far as I can tell, this definition is not consistent with what is really required.

The implementation of `createSign` in node-sshpk library is based on a [call](https://github.com/joyent/node-sshpk/blob/1da5a874f62ecc4a7593606020a230ba6a885aa3/lib/private-key.js#L130
) to the function `createSign()` in the native `crypto` module.

According to the [documentation](https://nodejs.org/api/crypto.html#cryptocreatesignalgorithm-options
), the function `createSign()` in the native crypto module accepts a list of string parameters as argument.
This list is available through a call to the [function](https://nodejs.org/api/crypto.html#cryptogethashes
) `crypto.getHashes()` 

On Node.js v17.3.1, this is what I get
```node
> crypto.getHashes()
[
  'RSA-MD4',
  'RSA-MD5',
  'RSA-MDC2',
  'RSA-RIPEMD160',
  'RSA-SHA1',
  'RSA-SHA1-2',
  'RSA-SHA224',
  'RSA-SHA256',
  'RSA-SHA3-224',
  'RSA-SHA3-256',
  'RSA-SHA3-384',
  'RSA-SHA3-512',
  'RSA-SHA384',
  'RSA-SHA512',
  'RSA-SHA512/224',
  'RSA-SHA512/256',
  'RSA-SM3',
  'blake2b512',
  'blake2s256',
  'id-rsassa-pkcs1-v1_5-with-sha3-224',
  'id-rsassa-pkcs1-v1_5-with-sha3-256',
  'id-rsassa-pkcs1-v1_5-with-sha3-384',
  'id-rsassa-pkcs1-v1_5-with-sha3-512',
  'md4',
  'md4WithRSAEncryption',
  'md5',
  'md5-sha1',
  'md5WithRSAEncryption',
  'mdc2',
  'mdc2WithRSA',
  'ripemd',
  'ripemd160',
  'ripemd160WithRSA',
  'rmd160',
  'sha1',
  'sha1WithRSAEncryption',
  'sha224',
  'sha224WithRSAEncryption',
  'sha256',
  'sha256WithRSAEncryption',
  'sha3-224',
  'sha3-256',
  'sha3-384',
  'sha3-512',
  'sha384',
  'sha384WithRSAEncryption',
  'sha512',
  'sha512-224',
  'sha512-224WithRSAEncryption',
  'sha512-256',
  'sha512-256WithRSAEncryption',
  'sha512WithRSAEncryption',
  'shake128',
  'shake256',
  'sm3',
  'sm3WithRSAEncryption',
  'ssl3-md5',
  'ssl3-sha1',
  'whirlpool'
]
```

Shouldn't `@types/sshpk.PrivateKey.createSign()` require a type that gathers such list, or at least the[ uppercase-only items](https://github.com/joyent/node-sshpk/blob/1da5a874f62ecc4a7593606020a230ba6a885aa3/lib/private-key.js#L144) in this list?

[a comment by an automated account is not shown]

**Comment by CoalZombik** [comment]
You're right, this type is definitely odd. I made a typo mistake, because it must be `AlgorithmHashType` at least, but this type don't reflect all supported variants too.

Function `crypto.getHashes()` returns same array in node 12 and node 17, therefore I think it is ok to specify type as list of constant strings.

While I checking this issue, I found another typo in `FingerprintParseOptions`.

I created an fork branch with changes for both errors: https://github.com/DefinitelyTyped/DefinitelyTyped/compare/master...CoalZombik:sshpk-hashalgo
If these changes is valid for you, I will create a pull request.

**Reply by CoalZombik** [reply]
Friendly ping @OlivierMartineau
